### PR TITLE
Refactor BigQueryServices to support extract and query jobs

### PR DESCRIPTION
**Pull request by peihe**

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java [hunk 1]
```diff
@@ -1168,6 +1170,10 @@ public BigQuerySink(
       // The maximum number of retry load jobs.
       private static final int MAX_RETRY_LOAD_JOBS = 3;
 
+      // The maximum number of retries to poll the status of a load job.
+      // It sets to {@code Integer.MAX_VALUE} to block until the BigQuery load job finishes.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java]: s/load/job/

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java [hunk 2]
```diff
@@ -1669,6 +1676,29 @@ public PDone apply(PCollection<TableRow> input) {
...
+  /**
+   * Status of a BigQuery job or request.
+   */
+  enum Status {
+    SUCCEEDED,
+    FAILED,
+    UNKNOWN,
+  }
+
+  private static Status parseStatus(Job job) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java]: @Nullable

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java [hunk 3]
```diff
@@ -46,92 +47,126 @@
 public class BigQueryServicesImpl implements BigQueryServices {
 
   // The maximum number of attempts to execute a load job RPC.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java]: s/load job/BigQuery/?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java (hunk 1 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java (hunk 2 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java (hunk 3 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java [hunk 4]
```diff
@@ -149,44 +184,47 @@ void startLoadJob(
...
           // The job is not DONE, wait longer and retry.
         } catch (IOException e) {
           // ignore and retry
           LOG.warn("Ignore the error and retry polling job status.", e);
         }
       } while (nextBackOff(sleeper, backoff));
-      LOG.warn("Unable to poll job status: {}, aborting after {} retries.",
-          jobId, MAX_LOAD_JOB_POLL_RETRIES);
-      return Status.UNKNOWN;
+      LOG.warn("Unable to poll job status: {}, aborting after reached max retries.", jobId);
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java]: error?

Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java]: and try to make it line up better with the existing BigQuery messages. When users hit them, it needs to be very clear what happened.

**Comment by dhalperi** [comment]
LGTM. Do you want to fix the error messages here or in the followup PR?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java (hunk 4 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/BigQueryServicesImpl.java]: I want to let the client of this class (BigQueryIO) to decide whether it is an error and warning, and decide the messages.

**Pull request closed without merging**